### PR TITLE
Line-by-line profiling

### PR DESCRIPTION
**Pull request by DamianX**
INTRODUCING PROFILING TECHNIQUES FROM THE 90s
Requires #16528
Ports https://github.com/tgstation/tgstation/pull/32116

### How to

- Enable debug verbs
- Profiling tab ->start line profiling verb
- Run the thing that has `PROFILE_TICK` sprinkled all over
- Show line profiling verb
- When you're done use the stop line profiling verb to clear the results

Example usage: https://gist.github.com/DamianX/b7d13151e077b915bc6a7ceaaee4a909
Example result:

### CODE REVIEW

File: __DEFINES/profile.dm [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+				PROFILE_ITEM[PROFILE_ITEM_COUNT] = 0;\
+			};\
+			PROFILE_ITEM[PROFILE_ITEM_TIME] += TICK_DELTA_TO_MS(PROFILE_TICK_USAGE_REAL-PROFILE_TIME);\
+			var/PROFILE_INCR_AMOUNT = min(1, 2**round(PROFILE_ITEM[PROFILE_ITEM_COUNT]/SHORT_REAL_LIMIT));\
+			if (prob(100/PROFILE_INCR_AMOUNT)) {\
+				PROFILE_ITEM[PROFILE_ITEM_COUNT] += PROFILE_INCR_AMOUNT;\
+			};\
+		};\
+		PROFILE_SET;\
+	};
```
Comment (N3X15) [inline __DEFINES/profile.dm:25]: `*scream`

Comment (DamianX) [reply __DEFINES/profile.dm:25]: huh?

**Comment by PJB3005** [comment]
@clusterfack

**Comment by clusterfack** [comment]
Wow spooky, if it does what it looks like it would replace chunking up the proc into separate procs to profile it

**Comment by gbasood** [comment]
I thought about doing something kinda like this once, but this is much more in depth
@clusterfack this also has the benefit of being basically world.log debug output except we can toggle it at runtime, so we can keep it in for things like lighting and then toggle it on when shit's getting weird

File: __DEFINES/profile.dm [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#define PROFILE_START ;PROFILE_STORE = list();PROFILE_SET;
```
Comment (PJB3005) [inline __DEFINES/profile.dm:1]: These have no reason reason to be defines.

Comment (DamianX) [reply __DEFINES/profile.dm:1]: Eh, I like them being defines, makes what you're doing more explicit. Can we keep it as is?

Comment (PJB3005) [reply __DEFINES/profile.dm:1]: Fine I guess.

**Pull request merged**